File: frontend/src/modules/story-bible/domain/story-bible.ts
export interface Character {
  id: string;
  name: string;
  role: "protagonist";
  desire: string;
  hiddenFeeling: string;
}

export interface WorldEntry {
  id: string;
  kind: "place" | "object" | "rule";
  title: string;
  description: string;
}

export interface WorldEntryDraftValue {
  kind: WorldEntry["kind"];
  title: string;
  description: string;
}

export interface WorldEntryDraftErrors {
  title?: string;
  description?: string;
}

export interface StoryBible {
  projectId: string;
  characters: Character[];
  worldEntries: WorldEntry[];
}

export interface SceneContextReference {
  characterIds: string[];
  worldEntryIds: string[];
}

export interface SceneContext {
  characters: Character[];
  worldEntries: WorldEntry[];
}

export function createInitialStoryBible(
  projectId: string,
  protagonistNames: [string, string],
): StoryBible {
  return {
    projectId,
    characters: protagonistNames.map((name, index) => ({
      id: `${projectId}-character-${index + 1}`,
      name,
      role: "protagonist" as const,
      desire:
        index === 0
          ? "상대에게 흔들리지 않고 자신의 선택을 지키고 싶다."
          : "과거의 오해를 풀고 다시 신뢰받고 싶다.",
      hiddenFeeling:
        index === 0 ? "여전히 상대의 진심을 확인하고 싶다." : "이번에는 먼저 놓치고 싶지 않다.",
    })),
    worldEntries: [
      {
        id: `${projectId}-world-1`,
        kind: "place",
        title: "비가 그친 온실",
        description:
          "두 사람이 과거에 마지막으로 만났던 장소. 젖은 흙과 오래된 장미 향이 남아 있다.",
      },
    ],
  };
}

export function selectSceneContext(
  bible: StoryBible,
  reference: SceneContextReference,
): SceneContext {
  return {
    characters: bible.characters.filter(({ id }) => reference.characterIds.includes(id)),
    worldEntries: bible.worldEntries.filter(({ id }) => reference.worldEntryIds.includes(id)),
  };
}

export function validateWorldEntryDraft(value: WorldEntryDraftValue): {
  value?: WorldEntryDraftValue;
  errors: WorldEntryDraftErrors;
} {
  const normalized = {
    ...value,
    title: value.title.trim(),
    description: value.description.trim(),
  };
  const errors: WorldEntryDraftErrors = {};

  if (!normalized.title) errors.title = "제목을 입력해 주세요.";
  if (!normalized.description) errors.description = "설명을 입력해 주세요.";

  return Object.keys(errors).length > 0 ? { errors } : { value: normalized, errors };
}

File: backend/apps/story_bible/repository/story_bible.py
import fcntl
import json
import os
import tempfile
from contextlib import contextmanager, suppress
from pathlib import Path
from typing import Any

from apps.story_bible.service.story_bible import (
    Character,
    StoryBible,
    StoryBibleNotFoundError,
    StoryBiblePersistenceError,
    StoryBibleRevisionConflictError,
    StoryBibleSnapshot,
    WorldEntry,
)

_SCHEMA_VERSION = 1


class FileStoryBibleRepository:
    def __init__(self, data_root: Path) -> None:
        self._data_root = data_root.resolve()

    def get(self, project_id: str) -> StoryBibleSnapshot:
        return self._read(self._story_bible_path(project_id), project_id)

    def replace(
        self,
        project_id: str,
        expected_revision: int,
        story_bible: StoryBible,
    ) -> StoryBibleSnapshot:
        path = self._story_bible_path(project_id)
        if story_bible.project_id != project_id:
            raise StoryBiblePersistenceError("Story Bible project does not match its path")
        if not path.parent.is_dir():
            raise StoryBibleNotFoundError

        with self._project_lock(path):
            current = self._read(path, project_id)
            if current.revision != expected_revision:
                raise StoryBibleRevisionConflictError
            replacement = StoryBibleSnapshot(
                story_bible=story_bible,
                revision=current.revision + 1,
            )
            self._atomic_write(path, replacement)
            return replacement

    def _story_bible_path(self, project_id: str) -> Path:
        projects_root = (self._data_root / "projects").resolve()
        candidate = (projects_root / project_id / "story-bible.json").resolve()
        if (
            not project_id
            or Path(project_id).name != project_id
            or not projects_root.is_relative_to(self._data_root)
            or not candidate.is_relative_to(self._data_root)
            or not candidate.is_relative_to(projects_root)
        ):
            raise StoryBiblePersistenceError("Project path escapes the configured data root")
        return candidate

    @contextmanager
    def _project_lock(self, story_bible_path: Path):
        lock_path = story_bible_path.with_suffix(".lock")
        try:
            with lock_path.open("a+b") as lock_file:
                fcntl.flock(lock_file.fileno(), fcntl.LOCK_EX)
                try:
                    yield
                finally:
                    fcntl.flock(lock_file.fileno(), fcntl.LOCK_UN)
        except (StoryBibleNotFoundError, StoryBibleRevisionConflictError):
            raise
        except OSError as error:
            raise StoryBiblePersistenceError("Could not lock Story Bible") from error

    @staticmethod
    def _read(path: Path, project_id: str) -> StoryBibleSnapshot:
        try:
            with path.open(encoding="utf-8") as source:
                document = json.load(source)
            return _decode_snapshot(document, project_id)
        except FileNotFoundError as error:
            raise StoryBibleNotFoundError from error
        except StoryBiblePersistenceError:
            raise
        except (
            OSError,
            UnicodeError,
            json.JSONDecodeError,
            TypeError,
            ValueError,
            KeyError,
        ) as error:
            raise StoryBiblePersistenceError("Could not read Story Bible") from error

    @staticmethod
    def _atomic_write(path: Path, snapshot: StoryBibleSnapshot) -> None:
        temp_path: Path | None = None
        try:
            with tempfile.NamedTemporaryFile(
                mode="w",
                encoding="utf-8",
                dir=path.parent,
                prefix=".story-bible.",
                suffix=".tmp",
                delete=False,
            ) as temporary:
                temp_path = Path(temporary.name)
                json.dump(
                    _encode_snapshot(snapshot),
                    temporary,
                    ensure_ascii=False,
                    separators=(",", ":"),
                )
                temporary.flush()
                os.fsync(temporary.fileno())
            os.replace(temp_path, path)
            temp_path = None
        except (OSError, TypeError, ValueError) as error:
            raise StoryBiblePersistenceError("Could not persist Story Bible") from error
        finally:
            if temp_path is not None:
                with suppress(OSError):
                    temp_path.unlink(missing_ok=True)


def _encode_snapshot(snapshot: StoryBibleSnapshot) -> dict[str, Any]:
    return {
        "schemaVersion": _SCHEMA_VERSION,
        "storyBibleRevision": snapshot.revision,
        "storyBible": {
            "projectId": snapshot.story_bible.project_id,
            "characters": [
                {
                    "id": character.id,
                    "name": character.name,
                    "role": character.role,
                    "desire": character.desire,
                    "hiddenFeeling": character.hidden_feeling,
                }
                for character in snapshot.story_bible.characters
            ],
            "worldEntries": [
                {
                    "id": entry.id,
                    "kind": entry.kind,
                    "title": entry.title,
                    "description": entry.description,
                }
                for entry in snapshot.story_bible.world_entries
            ],
        },
    }


def _decode_snapshot(document: object, project_id: str) -> StoryBibleSnapshot:
    envelope = _object_with_keys(document, {"schemaVersion", "storyBibleRevision", "storyBible"})
    if _integer(envelope["schemaVersion"]) != _SCHEMA_VERSION:
        raise StoryBiblePersistenceError("Unsupported Story Bible schema version")
    revision = _integer(envelope["storyBibleRevision"])
    if revision < 1:
        raise StoryBiblePersistenceError("Story Bible revision must be positive")

    payload = _object_with_keys(envelope["storyBible"], {"projectId", "characters", "worldEntries"})
    stored_project_id = _nonempty_string(payload["projectId"])
    if stored_project_id != project_id:
        raise StoryBiblePersistenceError("Story Bible project does not match its path")

    characters_value = _list(payload["characters"])
    entries_value = _list(payload["worldEntries"])
    characters = tuple(_decode_character(value) for value in characters_value)
    entries = tuple(_decode_world_entry(value) for value in entries_value)
    return StoryBibleSnapshot(
        story_bible=StoryBible(
            project_id=stored_project_id,
            characters=characters,
            world_entries=entries,
        ),
        revision=revision,
    )


def _decode_character(value: object) -> Character:
    item = _object_with_keys(value, {"id", "name", "role", "desire", "hiddenFeeling"})
    role = _string(item["role"])
    if role != "protagonist":
        raise StoryBiblePersistenceError("Invalid character role")
    return Character(
        id=_nonempty_string(item["id"]),
        name=_nonempty_string(item["name"]),
        role="protagonist",
        desire=_string(item["desire"]),
        hidden_feeling=_string(item["hiddenFeeling"]),
    )


def _decode_world_entry(value: object) -> WorldEntry:
    item = _object_with_keys(value, {"id", "kind", "title", "description"})
    kind = _string(item["kind"])
    if kind not in {"place", "object", "rule"}:
        raise StoryBiblePersistenceError("Invalid world entry kind")
    title = _nonempty_string(item["title"], strip=True)
    description = _nonempty_string(item["description"], strip=True)
    return WorldEntry(
        id=_nonempty_string(item["id"]),
        kind=kind,
        title=title,
        description=description,
    )


def _object_with_keys(value: object, keys: set[str]) -> dict[str, object]:
    if (
        not isinstance(value, dict)
        or set(value) != keys
        or not all(isinstance(key, str) for key in value)
    ):
        raise StoryBiblePersistenceError("Invalid Story Bible object")
    return value


def _list(value: object) -> list[object]:
    if not isinstance(value, list):
        raise StoryBiblePersistenceError("Invalid Story Bible list")
    return value


def _string(value: object) -> str:
    if not isinstance(value, str):
        raise StoryBiblePersistenceError("Invalid Story Bible string")
    return value


def _nonempty_string(value: object, *, strip: bool = False) -> str:
    result = _string(value)
    normalized = result.strip() if strip else result
    if not normalized:
        raise StoryBiblePersistenceError("Story Bible string must not be empty")
    return normalized


def _integer(value: object) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise StoryBiblePersistenceError("Invalid Story Bible integer")
    return value
